compiler/noirc_evaluator/src/ssa/opt/pure.rs

```diff
@@ -228,63 +234,68 @@ fn analyze_call_graph(
-
-        while let Some(index) = dfs.next(graph) {
-            let id = indices_to_ids[&index];
-            let mut purity = starting_purities[&id];
-
-            for neighbor_index in graph.neighbors(index) {
-                let neighbor = indices_to_ids[&neighbor_index];
-
-                let neighbor_purity = finished_purities.get(&neighbor).copied().unwrap_or({
-                    // The dependent function isn't finished yet. Since we're following
-                    // calls in a DFS, this means there are mutually recursive functions.
-                    // We could handle these but would need a different, much slower algorithm
-                    // to detect strongly connected components. Instead, since this should be
-                    // a rare case, we bail and assume impure for now.
-                    if neighbor == id {
-                        // If the recursive call is to the same function we can ignore it,
-                        // unless it's a brillig function in which case we need to set it to impure
-                        // as it could recurse inifinitely, and that's a side-effect
-                        if functions[&id].runtime().is_brillig() { Purity::Impure } else { purity }
-                    } else {
-                        Purity::Impure
+    let mut finished = HashMap::default();
+
+    // Map FunctionId -> SCC index for quick lookup
+    let mut func_to_scc = HashMap::default();
+    for (i, scc) in sccs.iter().enumerate() {
+        for &func in scc {
+            // Each function belongs to exactly one SCC by definition of SCCs.
+            // Therefore inserting into func_to_scc here is safe, and there will
+            // be no overwrites.
+            func_to_scc.insert(func, i);
+        }
+    }
+
+    // Track SCC purity
+    let mut scc_purities: Vec<Purity> = sccs
+        .iter()
+        .map(|scc| scc.iter().map(|f| starting_purities[f]).fold(Purity::Pure, |a, b| a.unify(b)))
+        .collect();
+
+    // Iteratively propagate purity between SCCs until convergence
+    let mut changed = true;
+    while changed {
+        changed = false;
+
+        for (i, scc) in sccs.iter().enumerate() {
+            let mut combined_purity = scc_purities[i];
+
+            // Look at neighbors outside the SCC
+            for &func in scc {
+                let idx = call_graph.ids_to_indices()[&func];
+                for neighbor_idx in call_graph.graph().neighbors(idx) {
+                    let neighbor = call_graph.indices_to_ids()[&neighbor_idx];
+                    let neighbor_scc = func_to_scc[&neighbor];
+                    if neighbor_scc != i {
+                        combined_purity = combined_purity.unify(scc_purities[neighbor_scc]);
                     }
-                });
-                purity = purity.unify(neighbor_purity);
+                }
+
+                // If this function is recursive AND a Brillig function, mark as PureWithPredicate
```

Comment:
We could explain why we do this (is it because it could run forever?)

Reply:
Nevermind, I was looking it commit by commit but now I see it's done in a next commit.